meter/shelly/connection.go

```diff
@@ -3,103 +3,104 @@ package shelly
 // Connection is the Shelly connection
 type Connection struct {
-	*request.Helper
-	uri     string
-	channel int
-	gen     int    // Shelly api generation
 	model   string // Shelly device type
 	profile string // Shelly device profile
+	Cache   time.Duration
+	Gen
 }
 
 // NewConnection creates a new Shelly device connection.
-func NewConnection(uri, user, password string, channel int) (*Connection, error) {
+func NewConnection(uri, user, password string, channel int, cache time.Duration) (*Connection, error) {
 	if uri == "" {
 		return nil, errors.New("missing uri")
 	}
 
 	for _, suffix := range []string{"/", "/rcp", "/shelly"} {
 		uri = strings.TrimSuffix(uri, suffix)
 	}
+	uri = util.DefaultScheme(uri, "http")
 
 	log := util.NewLogger("shelly")
 	client := request.NewHelper(log)
 
 	// Shelly Gen1 and Gen2 families expose the /shelly endpoint
 	var resp DeviceInfo
-	if err := client.GetJSON(fmt.Sprintf("%s/shelly", util.DefaultScheme(uri, "http")), &resp); err != nil {
+	if err := client.GetJSON(fmt.Sprintf("%s/shelly", uri), &resp); err != nil {
 		return nil, err
 	}
 
 	if (resp.Auth || resp.AuthEn) && (user == "" || password == "") {
 		return nil, fmt.Errorf("%s (%s) missing user/password", resp.Model, resp.Mac)
 	}
 
-	conn := &Connection{
-		Helper:  client,
-		uri:     util.DefaultScheme(uri, "http"),
-		channel: channel,
-		gen:     resp.Gen,
-		model:   strings.Split(resp.Type+resp.Model, "-")[0],
-		profile: resp.Profile,
-	}
-
-	conn.Client.Transport = request.NewTripper(log, transport.Insecure())
-
 	// Set default profile to "monophase" if not provided
 	if resp.Profile == "" {
 		resp.Profile = "monophase"
 	}
 
-	switch conn.gen {
-	case 0, 1:
+	model := strings.Split(resp.Type+resp.Model, "-")[0]
+
+	client.Transport = request.NewTripper(log, transport.Insecure())
+
+	var gen Gen
+	if resp.Gen < 2 {
 		// Shelly GEN 1 API
 		// https://shelly-api-docs.shelly.cloud/gen1/#shelly-family-overview
 		if user != "" {
 			log.Redact(transport.BasicAuthHeader(user, password))
-			conn.Client.Transport = transport.BasicAuth(user, password, conn.Client.Transport)
+			client.Transport = transport.BasicAuth(user, password, client.Transport)
 		}
-	case 2, 3:
+		gen = &gen1{
+			Helper:  client,
+			uri:     uri,
+			channel: channel,
+			model:   model,
+			status:  util.NewCacheable[Gen1Status](),
+		}
+	}
+	if resp.Gen > 1 {
 		// Shelly GEN 2+ API
 		// https://shelly-api-docs.shelly.cloud/gen2/
-		conn.uri += "/rpc"
-		if user != "" {
-			conn.Client.Transport = digest.NewTransport(user, password, conn.Client.Transport)
+		gen = &gen2{
+			Helper:       client,
+			uri:          uri,
+			channel:      channel,
+			model:        model,
+			profile:      resp.Profile,
+			switchstatus: util.NewCacheable[Gen2SwitchStatus](),
+			em1status:    util.NewCacheable[Gen2EM1Status](),
+			em1data:      util.NewCacheable[Gen2EM1Data](),
+			emstatus:     util.NewCacheable[Gen2EMStatus](),
+			emdata:       util.NewCacheable[Gen2EMData](),
```

Comment:
Replace with

    util.ResettableChached()

...wobei ich mich frage ob es wirklich soviele Caches braucht?

Reply:
bzw. das gehört dann einfach in 

   newGen2()